enterprise/server/remote_execution/snaploader/snaploader.go

```diff
@@ -314,23 +342,81 @@ func (l *FileCacheLoader) CacheSnapshot(ctx context.Context, key *fcpb.SnapshotK
 	return &Snapshot{key: key, manifest: manifest}, nil
 }
 
+func (l *FileCacheLoader) UnpackContainerImage(ctx context.Context, imageRef, imageExt4Path string, outDir string, chunkSize int64) (*ChunkedFile, error) {
+	ctx, span := tracing.StartSpan(ctx)
+	defer span.End()
+
+	gid, err := groupID(ctx, l.env)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: use a real Action as the cache key.
+	hash := hashStrings(gid, "containerfs-chunked", imageRef, fmt.Sprint(chunkSize))
+	metadataKey := &repb.FileNode{Digest: &repb.Digest{Hash: hash}}
+	md, err := l.getChunkedFileMetadata(ctx, metadataKey)
+	if err != nil && !status.IsNotFoundError(err) {
+		return nil, err
+	}
+	if md != nil {
+		// Chunked file is available in cache; just unpack it.
+		return l.unpackCOW(ctx, md, outDir)
+	}
+	// Chunked file is not available in cache; convert the EXT4 image to a
+	// ChunkedFile then add it to cache.
+	// TODO(bduffany): single-flight this.
+	start := time.Now()
+	cow, err := blockio.ConvertFileToCOW(imageExt4Path, chunkSize, outDir)
+	if err != nil {
+		return nil, status.WrapError(err, "convert image to COW")
+	}
+	// Add the COW to cache. This will also compute chunk digests.
+	cf := &ChunkedFile{COWStore: cow}
+	if _, err := l.cacheCOW(ctx, "rootfs.ext4", cf); err != nil {
+		return nil, err
+	}
+	log.Debugf("Converted containerfs to COW in %s", time.Since(start))
+	return cf, nil
+}
```

Comment:
I just realized that we can simplify all of this if we make all fields optional in `Snapshot` (i.e. don't require a memory / vmstate file). Then, the containerfs artifact can just be a snapshot and reuse most of the existing code for getting snapshots from cache, validating snapshots etc. wdyt?

Reply:
Yeah I like reusing the existing architecture. It's confusing that we have a couple functions that are very similar but slightly different